diff --git a/dist/index.d.ts b/dist/index.d.ts
index c9cc51d8b715b6f220d67574b8c674247d032926..9b4c2e6b096a0a139792d3fd4ce89c2f15bfc51e 100644
--- a/dist/index.d.ts
+++ b/dist/index.d.ts
@@ -1,11 +1,15 @@
 import { WithUnknown } from '@xsai/shared';
-import { ChatOptions, CompletionStep, PostToolCall, PrepareStep, PreToolCall, StopCondition, Message, Usage, FinishReason, AssistantMessage, ChatCompletionUsage, CompletionToolCall, CompletionToolResult } from '@xsai/shared-chat';
+import { ChatOptions, CompletionStep, OnToolCallFinishCallback, OnToolCallStartCallback, PostToolCall, PrepareStep, PreToolCall, RepairToolCallFunction, StopCondition, Message, Usage, FinishReason, AssistantMessage, ChatCompletionUsage, CompletionToolCall, CompletionToolResult } from '@xsai/shared-chat';
 
 interface GenerateTextOptions extends ChatOptions {
     onStepFinish?: (step: CompletionStep<true>) => Promise<unknown> | unknown;
+    captureToolErrors?: boolean;
+    onToolCallFinish?: OnToolCallFinishCallback;
+    onToolCallStart?: OnToolCallStartCallback;
     postToolCall?: PostToolCall;
     prepareStep?: PrepareStep;
     preToolCall?: PreToolCall;
+    repairToolCall?: RepairToolCallFunction;
     /** @internal */
     steps?: CompletionStep<true>[];
     /** @default `stepCountAtLeast(1)` */
diff --git a/dist/index.js b/dist/index.js
index d4fdab8b13f83fc2eb55eca0d15919133347393e..334428415f242fc6b11c874e988bc98187c3eadc 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -55,9 +55,13 @@ const rawGenerateText = async (options) => {
       const results = await Promise.all(
         msgToolCalls.map(async (toolCall) => executeTool({
           abortSignal: options.abortSignal,
+          captureToolErrors: options.captureToolErrors,
           messages,
+          onToolCallFinish: options.onToolCallFinish,
+          onToolCallStart: options.onToolCallStart,
           postToolCall: options.postToolCall,
           preToolCall: options.preToolCall,
+          repairToolCall: options.repairToolCall,
           toolCall,
           tools: options.tools
         }))
