// @moeru/eventa@1.0.0-alpha.14/dist/context-ex8urwfs.js
import { __toESM } from "../../../../../../../_virtual/_rolldown/runtime.js";
import { require_picomatch } from "../../../../../picomatch@4.0.3/node_modules/picomatch/index.js";
import { customAlphabet } from "../../../../../nanoid@5.1.6/node_modules/nanoid/index.browser.js";

//#region ../../node_modules/.pnpm/@moeru+eventa@1.0.0-alpha.14_electron@39.7.0/node_modules/@moeru/eventa/dist/context-ex8urwfs.mjs
var import_picomatch = /* @__PURE__ */ __toESM(require_picomatch(), 1);
function nanoid() {
	return customAlphabet("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz", 16)();
}
let EventaType = /* @__PURE__ */ function(EventaType$1) {
	EventaType$1["Event"] = "event";
	EventaType$1["MatchExpression"] = "matchExpression";
	return EventaType$1;
}({});
let EventaFlowDirection = /* @__PURE__ */ function(EventaFlowDirection$1) {
	EventaFlowDirection$1["Inbound"] = "inbound";
	EventaFlowDirection$1["Outbound"] = "outbound";
	return EventaFlowDirection$1;
}({});
function defineInboundEventa(id) {
	return {
		...defineEventa(id),
		_flowDirection: EventaFlowDirection.Inbound
	};
}
function defineOutboundEventa(id) {
	return {
		...defineEventa(id),
		_flowDirection: EventaFlowDirection.Outbound
	};
}
function defineEventa(id) {
	if (!id) id = nanoid();
	return {
		id,
		type: EventaType.Event
	};
}
function and(...matchExpression) {
	return {
		id: nanoid(),
		type: EventaType.MatchExpression,
		matcher: (event) => {
			return matchExpression.every((m) => m.matcher ? m.matcher(event) : false);
		}
	};
}
function matchBy(matchExpressionPossibleValues, inverted) {
	const id = nanoid();
	let matcher = () => false;
	if (typeof matchExpressionPossibleValues === "string") matcher = (eventa) => {
		return (0, import_picomatch.default)(matchExpressionPossibleValues)(eventa.id);
	};
	else if (typeof matchExpressionPossibleValues === "object") {
		if ("ids" in matchExpressionPossibleValues) matcher = (event) => {
			if (inverted) return !matchExpressionPossibleValues.ids.includes(event.id);
			return matchExpressionPossibleValues.ids.includes(event.id);
		};
		else if ("eventa" in matchExpressionPossibleValues) matcher = (event) => {
			if (inverted) return !matchExpressionPossibleValues.eventa.some((e) => e.id === event.id);
			return matchExpressionPossibleValues.eventa.some((e) => e.id === event.id);
		};
		else if ("types" in matchExpressionPossibleValues) matcher = (event) => {
			if (typeof event.type === "undefined") return false;
			if (inverted) return !matchExpressionPossibleValues.types.includes(event.type);
			return matchExpressionPossibleValues.types.includes(event.type);
		};
	} else if (matchExpressionPossibleValues instanceof RegExp) matcher = (event) => {
		if (inverted) return !matchExpressionPossibleValues.test(event.id);
		return matchExpressionPossibleValues.test(event.id);
	};
	else if (typeof matchExpressionPossibleValues === "function") matcher = matchExpressionPossibleValues;
	return {
		id,
		type: EventaType.MatchExpression,
		matcher
	};
}
function createContext(props = {}) {
	const listeners = /* @__PURE__ */ new Map();
	const onceListeners = /* @__PURE__ */ new Map();
	const matchExpressions = /* @__PURE__ */ new Map();
	const matchExpressionListeners = /* @__PURE__ */ new Map();
	const matchExpressionOnceListeners = /* @__PURE__ */ new Map();
	const hooks = props.adapter?.(emit).hooks;
	function emit(event, payload, options) {
		const emittingPayload = {
			...event,
			body: payload
		};
		for (const listener of listeners.get(event.id) || []) {
			listener(emittingPayload, options);
			hooks?.onReceived?.(event.id, emittingPayload);
		}
		for (const onceListener of onceListeners.get(event.id) || []) {
			onceListener(emittingPayload, options);
			hooks?.onReceived?.(event.id, emittingPayload);
			onceListeners.get(event.id)?.delete(onceListener);
		}
		for (const matchExpression of matchExpressions.values()) if (matchExpression.matcher) {
			if (!matchExpression.matcher(emittingPayload)) continue;
			for (const listener of matchExpressionListeners.get(matchExpression.id) || []) {
				listener(emittingPayload, options);
				hooks?.onReceived?.(matchExpression.id, emittingPayload);
			}
			for (const onceListener of matchExpressionOnceListeners.get(matchExpression.id) || []) {
				onceListener(emittingPayload, options);
				hooks?.onReceived?.(matchExpression.id, emittingPayload);
				matchExpressionOnceListeners.get(matchExpression.id)?.delete(onceListener);
			}
		}
		hooks?.onSent(event.id, emittingPayload, options);
	}
	return {
		get listeners() {
			return listeners;
		},
		get onceListeners() {
			return onceListeners;
		},
		emit,
		on(eventOrMatchExpression, handler) {
			if (eventOrMatchExpression.type === EventaType.Event) {
				const event = eventOrMatchExpression;
				if (!listeners.has(event.id)) listeners.set(event.id, /* @__PURE__ */ new Set());
				listeners.get(event.id)?.add(handler);
				return () => listeners.get(event.id)?.delete(handler);
			}
			if (eventOrMatchExpression.type === EventaType.MatchExpression) {
				const matchExpression = eventOrMatchExpression;
				if (!matchExpressions.has(matchExpression.id)) matchExpressions.set(matchExpression.id, matchExpression);
				if (!matchExpressionListeners.has(matchExpression.id)) matchExpressionListeners.set(matchExpression.id, /* @__PURE__ */ new Set());
				matchExpressionListeners.get(matchExpression.id)?.add(handler);
				return () => matchExpressionListeners.get(matchExpression.id)?.delete(handler);
			}
			return () => void 0;
		},
		once(eventOrMatchExpression, handler) {
			if (eventOrMatchExpression.type === EventaType.Event) {
				const event = eventOrMatchExpression;
				if (!onceListeners.has(event.id)) onceListeners.set(event.id, /* @__PURE__ */ new Set());
				onceListeners.get(event.id)?.add(handler);
				return () => onceListeners.get(event.id)?.delete(handler);
			}
			if (eventOrMatchExpression.type === EventaType.MatchExpression) {
				const matchExpression = eventOrMatchExpression;
				if (!matchExpressions.has(matchExpression.id)) matchExpressions.set(matchExpression.id, matchExpression);
				if (!matchExpressionListeners.has(matchExpression.id)) matchExpressionListeners.set(matchExpression.id, /* @__PURE__ */ new Set());
				matchExpressionOnceListeners.get(matchExpression.id)?.add(handler);
				return () => matchExpressionOnceListeners.get(matchExpression.id)?.delete(handler);
			}
			return () => void 0;
		},
		off(eventOrMatchExpression, handler) {
			switch (eventOrMatchExpression.type) {
				case EventaType.Event:
					if (handler !== void 0) {
						listeners.get(eventOrMatchExpression.id)?.delete(handler);
						onceListeners.get(eventOrMatchExpression.id)?.delete(handler);
						break;
					}
					listeners.delete(eventOrMatchExpression.id);
					onceListeners.delete(eventOrMatchExpression.id);
					break;
				case EventaType.MatchExpression:
					if (handler !== void 0) {
						matchExpressionListeners.get(eventOrMatchExpression.id)?.delete(handler);
						matchExpressionOnceListeners.get(eventOrMatchExpression.id)?.delete(handler);
						break;
					}
					matchExpressionListeners.delete(eventOrMatchExpression.id);
					matchExpressionOnceListeners.delete(eventOrMatchExpression.id);
					break;
			}
		}
	};
}

//#endregion
export { EventaFlowDirection, and, createContext, defineEventa, defineInboundEventa, defineOutboundEventa, matchBy, nanoid };
