// @moeru/eventa@1.0.0-alpha.14/dist/src-OOE3RB9u.js
import { defineEventa, nanoid } from "./context-ex8urwfs.js";

//#region ../../node_modules/.pnpm/@moeru+eventa@1.0.0-alpha.14_electron@39.7.0/node_modules/@moeru/eventa/dist/src-OOE3RB9u.mjs
/**
* Read the internal invoke configuration from context extensions, if present.
* This is used by defineInvoke to determine which events should abort inflight invokes.
*/
function getContextExtensionInvokeInternalConfig(ctx) {
	const extensions = ctx.extensions;
	if (!extensions || typeof extensions !== "object") return;
	const internal = extensions.__internal;
	if (!internal || typeof internal !== "object") return;
	const invoke = internal.invoke;
	if (!invoke || typeof invoke !== "object") return;
	return invoke;
}
let InvokeEventType = /* @__PURE__ */ function(InvokeEventType$1) {
	InvokeEventType$1[InvokeEventType$1["SendEvent"] = 0] = "SendEvent";
	InvokeEventType$1[InvokeEventType$1["SendEventError"] = 1] = "SendEventError";
	InvokeEventType$1[InvokeEventType$1["SendEventStreamEnd"] = 2] = "SendEventStreamEnd";
	InvokeEventType$1[InvokeEventType$1["SendEventAbort"] = 3] = "SendEventAbort";
	InvokeEventType$1[InvokeEventType$1["ReceiveEvent"] = 4] = "ReceiveEvent";
	InvokeEventType$1[InvokeEventType$1["ReceiveEventError"] = 5] = "ReceiveEventError";
	InvokeEventType$1[InvokeEventType$1["ReceiveEventStreamEnd"] = 6] = "ReceiveEventStreamEnd";
	return InvokeEventType$1;
}({});
function defineInvokeEventa(tag) {
	if (!tag) tag = nanoid();
	return {
		sendEvent: {
			...defineEventa(`${tag}-send`),
			invokeType: InvokeEventType.SendEvent
		},
		sendEventError: {
			...defineEventa(`${tag}-send-error`),
			invokeType: InvokeEventType.SendEventError
		},
		sendEventStreamEnd: {
			...defineEventa(`${tag}-send-stream-end`),
			invokeType: InvokeEventType.SendEventStreamEnd
		},
		sendEventAbort: {
			...defineEventa(`${tag}-send-abort`),
			invokeType: InvokeEventType.SendEventAbort
		},
		receiveEvent: {
			...defineEventa(`${tag}-receive`),
			invokeType: InvokeEventType.ReceiveEvent
		},
		receiveEventError: {
			...defineEventa(`${tag}-receive-error`),
			invokeType: InvokeEventType.ReceiveEventError
		},
		receiveEventStreamEnd: {
			...defineEventa(`${tag}-receive-stream-end`),
			invokeType: InvokeEventType.ReceiveEventStreamEnd
		}
	};
}
/**
* Checks if a value is an AsyncIterable.
*
* @param value
* @returns True if the value is an AsyncIterable.
*/
function isAsyncIterable(value) {
	return typeof value === "object" && value !== null && Symbol.asyncIterator in value;
}
/**
* Checks if an object is a ReadableStream.
*
* @link https://github.com/cloudflare/workerd/blob/88e8696ce7a5f8969a7e02a2dcfb6504c17c9e8d/src/cloudflare/internal/streaming-forms.ts#L3
* @param obj
* @returns True if the object looks like a ReadableStream.
*/
function isReadableStream(obj) {
	return !!(obj && typeof obj === "object" && "getReader" in obj && typeof obj.getReader === "function");
}
function createAbortError(reason) {
	if (reason instanceof Error && reason.name === "AbortError") return reason;
	if (typeof DOMException !== "undefined") try {
		return new DOMException(reason ? String(reason) : "Aborted", "AbortError");
	} catch {}
	const error = reason instanceof Error ? reason : new Error(reason ? String(reason) : "Aborted");
	error.name = "AbortError";
	return error;
}
function isAbortError(error) {
	return error instanceof Error && error.name === "AbortError";
}
/**
* Create a unary invoke function (client side).
*
* It supports unary or streaming requests, but returns a single response.
* Use `defineStreamInvoke` when you expect a stream of responses.
*
* If you want stream input, set `Req` to `ReadableStream<T>` or `AsyncIterable<T>`
* (or a union type like `T | ReadableStream<T>` for optional streaming).
*
* @example
* ```ts
* // 1) Define eventa once (shared by client/server)
* const events = defineInvokeEventa<{ id: string }, { name: string }>()
*
* // 2) Client: define invoke function
* const invoke = defineInvoke(clientCtx, events)
*
* // 3) Call
* const res = await invoke({ name: 'alice' })
* ```
*
* @example
* ```ts
* // Stream request -> unary response
* const events = defineInvokeEventa<number, ReadableStream<number>>()
*
* defineInvokeHandler(serverCtx, events, async (payload) => {
*   let sum = 0
*   for await (const value of payload) {
*     sum += value
*   }
*
*   return sum
* })
*
* const invoke = defineInvoke(clientCtx, events)
* const input = new ReadableStream<number>({
*   start(controller) {
*     controller.enqueue(1)
*     controller.enqueue(2)
*     controller.close()
*   },
* })
*
* const total = await invoke(input)
* ```
*
* @param ctx Event context on the caller/client side.
* @param event Invoke event definition created by `defineInvokeEventa`.
*/
function defineInvoke(ctx, event) {
	function _invoke(req, options) {
		return new Promise((resolve, reject) => {
			const invokeId = nanoid();
			const invokeReceiveEvent = defineEventa(`${event.receiveEvent.id}-${invokeId}`);
			const invokeReceiveEventError = defineEventa(`${event.receiveEventError.id}-${invokeId}`);
			const { signal, ...emitOptions } = options ?? {};
			let finished = false;
			const onAbort = () => {
				ctx.emit(event.sendEventAbort, {
					invokeId,
					content: signal?.reason
				}, emitOptions);
				finishReject(createAbortError(signal?.reason));
			};
			const invokeInternalConfig = getContextExtensionInvokeInternalConfig(ctx);
			const abortOffs = [];
			const onAbortEvent = (payload, eventOptions) => {
				const mappedError = invokeInternalConfig?.mapAbortError?.(payload, eventOptions);
				if (typeof mappedError !== "undefined") {
					finishReject(mappedError);
					return;
				}
				const body = payload?.body;
				finishReject(body && typeof body === "object" && "error" in body ? body.error : typeof body !== "undefined" ? body : payload);
			};
			const cleanup = () => {
				ctx.off(invokeReceiveEvent);
				ctx.off(invokeReceiveEventError);
				for (const off of abortOffs) off();
				if (signal) signal.removeEventListener("abort", onAbort);
			};
			const finishReject = (error) => {
				if (finished) return;
				finished = true;
				reject(error);
				cleanup();
			};
			const finishResolve = (value) => {
				if (finished) return;
				finished = true;
				resolve(value);
				cleanup();
			};
			ctx.on(invokeReceiveEvent, (payload) => {
				if (!payload.body) return;
				if (payload.body.invokeId !== invokeId) return;
				const { content } = payload.body;
				finishResolve(content);
			});
			ctx.on(invokeReceiveEventError, (payload) => {
				if (!payload.body) return;
				if (payload.body.invokeId !== invokeId) return;
				const { error } = payload.body.content;
				finishReject(error);
			});
			if (invokeInternalConfig?.abortOnEvents?.length) for (const eventOrMatch of invokeInternalConfig.abortOnEvents) abortOffs.push(ctx.on(eventOrMatch, onAbortEvent));
			if (signal) {
				if (signal.aborted) {
					onAbort();
					return;
				}
				signal.addEventListener("abort", onAbort, { once: true });
			}
			if (!isReadableStream(req) && !isAsyncIterable(req)) ctx.emit(event.sendEvent, {
				invokeId,
				content: req
			}, emitOptions);
			else {
				const sendChunk = (chunk) => {
					if (finished) return;
					ctx.emit(event.sendEvent, {
						invokeId,
						content: chunk,
						isReqStream: true
					}, emitOptions);
				};
				const sendEnd = () => {
					if (finished) return;
					ctx.emit(event.sendEventStreamEnd, {
						invokeId,
						content: void 0
					}, emitOptions);
				};
				const pump = async () => {
					try {
						for await (const chunk of req) {
							if (signal?.aborted) return;
							sendChunk(chunk);
						}
						sendEnd();
					} catch (error) {
						if (signal?.aborted) return;
						if (isAbortError(error)) {
							ctx.emit(event.sendEventAbort, {
								invokeId,
								content: error
							}, emitOptions);
							return;
						}
						ctx.emit(event.sendEventError, {
							invokeId,
							content: error
						}, emitOptions);
					}
				};
				pump();
			}
		});
	}
	return _invoke;
}

//#endregion
export { defineInvoke, defineInvokeEventa };
